### PR TITLE
event: Extract DispatcherBase interface

### PR DESCRIPTION
**Pull request by antoniovicente**
Commit Message:
event: Extract DispatcherBase interface

This new interface allows us to more safely build primitives like ScaledTimers which are themselves built on top of low-level 
event loop primitives like Timers which are themselves created via the dispatcher interface.
Additional Description:
Risk Level: n/a interface only change
Testing: n/a
Docs Changes: n/a
Release Notes: n/a
Issue #14401

### CODE REVIEW

**Review by akonradi — DISMISSED** [review]
Looks good. Have you tried refactoring any of the impl classes returned by the dispatcher to use DispatcherBase yet?

File: include/envoy/event/dispatcher.h [hunk 1]
```diff
@@ -52,12 +52,67 @@ using PostCb = std::function<void()>;
 using PostCbSharedPtr = std::shared_ptr<PostCb>;
 
 /**
- * Abstract event dispatching loop.
+ * Minimal interface to the disptching loop used to create low-level primitives. See Dispatcher
```
Comment (akonradi) [inline include/envoy/event/dispatcher.h:55]: nit: sp on `dispatching` is causing your format failure

**Comment by mattklein123** [comment]
This makes sense to me but this is small enough that do we want to just do this change in another change that actually needs it? Up to you all.

/wait

**Comment by antoniovicente** [comment]
> This makes sense to me but this is small enough that do we want to just do this change in another change that actually needs it? Up to you all.
> 
> /wait

@akonradi Thoughts?  Do you want to include this in a PR that adds a way to create scaled timers via the dispatcher?

**Review by mattklein123 — APPROVED** [review]

**Comment by akonradi** [comment]
Given that this is already approved, I say let's get it in now. No need to make the next PR bigger.

**Comment by antoniovicente** [comment]
Thanks Matt.

**Pull request merged**